### PR TITLE
[Feature] Edit extra pivot fields on a BelongsToMany relation, using the repeatable field

### PR DESCRIPTION
**Pull request by pxpm**
Last ref: #2860

Like the title describes and some issues have arised concerning this scenario we decided to give it a go and see where we can go with it.

So... the basic idea is to create a repeatable field that can accomodate the belongsToMany relation and any extra fields in that pivot table.

For it to work, you must have the pivot table and the relation must expose the pivot fields.

By example:

```php
//relation, the important part here is the ->withPivot() part.

public function products() {
     return $this->belongsToMany('App\Products', 'pivot_table_name')->withPivot(['attr1','attr2']);
}
```
```php
[                'label'     => 'BelongsToMany edit pivot fields.',
                'type'      => 'repeatable',
                'name'      => 'products', //the name of the BelongsToMany relation
                'fields' => [
                    [
                        'name' => 'products', //the name of the relation
                        'multiple' => false //this is mandatory.
                    ],
                    [
                            'name' => 'attr1',
                            'type' => 'text',
                            'label' => 'first extra attribute in pivot'
                     ],
                     [
                            'name' => 'attr2',
                            'type' => 'text',
                            'label' => 'second extra attribute in pivot'
                     ],
            ],
        ],
```

You are all set. Enjoy.

Notes: 
- Using `multiple => false` is mandatory as per relation definition belongsToMany allows multiple, but we only want one per line.

ping @tabacitu wanna give this a go ?

### CODE REVIEW

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 1]
```diff
@@ -112,6 +112,10 @@ public function syncPivot($model, $data)
             if (isset($field['pivot']) && $field['pivot']) {
                 $values = isset($data[$field['name']]) ? $data[$field['name']] : [];
 
+                //we override the values if the field is repeatable and pivot == true
+                if ($field['type'] == 'repeatable') {
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:116]: We just talked about this - you had the great idea to do things here NOT if the field type is repeatable, but if the:
- field has `pivot==true`
- field has `fields` array (exists and not empty)
- (maybe even) $field['fields'] contains a field with the same name as the parent field (the relationship)

Then we can basically be sure this is what the developer wants.

Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php]: Just did the changes.

**Comment by tabacitu** [comment]
Thanks a lot @pxpm ! Great work, I agree we should have this. But I'm going to postpone reviewing this new feature for now - fixing bugs takes priority this week and the next. New features can wait.

If anybody else needs this feature and can test it and comment / confirm it's working, that'd be great, it would definitely speed up the merge.

Cheers!

**Comment by rUriarteDev** [comment]
Wow i was expecting this a lot, i have tested and is working great @pxpm. Really nice feature, the only thing is missing when i try to delete the item is throwing integrity constraint violation, this of course is just because the item still has reference in the pivot table, so just need to detach before deleting. This can be managed through the destroy function, but will be nice to do integrate this in the PR what do you think. 
Cheers!

**Comment by pxpm** [comment]
@rUriarteDev hello :)

Thanks for giving this feature a try and providing valuable feedback. 

Do you have a working solution for the problem you described about deleting ? I am just wondering if it's not something that should be dealt at database level `onDelete('cascade')` or as a model event `deleting` ?

Best,
Pedro

**Comment by rUriarteDev** [comment]
Hi @pxpm  I'm not sure if it could be handled at the database level, maybe yes. I have done this simply in the destroy method (it could also be handled with an observer).
 ```
public function destroy($id)
    {
        $this->crud->hasAccessOrFail('delete');

        // get entry ID from Request (makes sure its the last ID for nested resources)
        $id = $this->crud->getCurrentEntryId() ?? $id;

        $plato= Plato::findOrFail($id);

        $plato->insumos()->detach();
        

        return $this->crud->delete($id);
    }
```

Note that the detach must be used before deleting the object $plato. If I do not do the detach before deleting the object, it still exists in the pivot table and for that reason it throws the error.

**Comment by pxpm** [comment]
Indeed, that is what I was talking about. 

There you used the destroy method, but you could have used the `deleting` model event (that you could use solo or in an observer class) but it might work also with the database automation. `onDelete('cascade')` (didn't tested, it's just a guess).

My thinking is that we cannot make it general because when model is deleting we don't know what to delete before deleting the model, that is something that usually developer need to account for. 

I think that about the deleting part, we should warn in the documentation that user needs to take care of that, but we are not able to provide an auto-solution. 

Maybe, in the best scenario, make the developer life easier creating some function like:

```php
public function deletePivotData() {
//delete, detach the pivot data.
}
```

And we can use that function by default in destroy function. 

That way developer could overwrite the `deletePivotData()` in the desired model, and we would run that function when deleting a crud entry before actually delete the item. 

Let me know if you think that this would be a proper solution for it. 

Best,
Pedro

**Comment by rUriarteDev** [comment]
For my part, I don't think it is a good practice to delete an object and continue to maintain a reference in the pivot table using an FK of an object that no longer exists.
Furthermore, I do not believe that the database allows such behavior. So I think removing the object should remove its references in the pivot table.
Maybe I'm not imagining the scenario in which that situation could occur.

**Comment by pxpm** [comment]
I agree with you. You should delete the pivot data related to the deleted entry for database consistency **unless** you are using soft-deletes in your model. 

I think it could be achieved with database fk `$table->integer('entry_id')->references('id')->on('entry_table')->onDelete('cascade')` to abstract the deletion to the database.

or using the model events:

```php
static::deleting(function($item) {
//delete the pivot
});
```

The way you solve it through `destroy()` is also a viable solution.

What I said is that deleting pivot data is developer responsibility, like you did or using any other method I just described.

Best,
Pedro

**Comment by rUriarteDev** [comment]
I agree with your point of view, it would only be necessary to clarify in the documentation that the programmer decides at what level (application, db, etc) to handle this.
From what I have tested it is working well so I hope it is available as soon as possible.
Thank you very much again.
Regards.
Rodrigo

**Comment by rUriarteDev** [comment]
Hi @tabacitu any news on this?
Regards.

Rodrigo

**Comment by kitzberger** [comment]
Just want to confirm that this patch seems to be working for our project with version `4.1.x`. We're using a `morphToMany` relation btw.

**Comment by pxpm** [comment]
@kitzberger thanks for giving it a try. 

I didn't close this yet, but ALL this and much more have been added into: #3410 

I will be closing this branch in a few. 

If you give that other PR a try let me know there, that one is the one that is going foward. 

Best,
Pedro

**Comment by pxpm** [comment]
This is merged in #3410 that we keep even with master until we merge it in 4.2

**Pull request closed without merging**